Use lazer info for CL scores set in lazer in pp calculation

Givikap120 opened the pull request:
Considering that lazer CL scores don't have ScoreV1 - they should use lazer info for true miss count. 
There's a problem with LTMs calculating as misses, making maps like Oshama Scramble heavily underweight, but there's nothing we can do with it, as we only have total amount of breaks.

Givikap120 committed:
pass "isLegacy" flag to score info

Givikap120 committed:
use lazer info for lazer CL scores

tsunyoku commented:
!diffcalc
RULESET=osu
OSU_A=https://github.com/ppy/osu/tree/pp-dev
OSU_B=https://github.com/ppy/osu/pull/33961
SCORE_PROCESSOR_A=https://github.com/tsunyoku/osu-queue-score-statistics/tree/add-hd-so-to-relevant-legacy-mods
SCORE_PROCESSOR_B=https://github.com/tsunyoku/osu-queue-score-statistics/tree/add-hd-so-to-relevant-legacy-mods

Givikap120 committed:
fix scorev2 line

tsunyoku commented:
!diffcalc
RULESET=osu
OSU_A=https://github.com/ppy/osu/tree/pp-dev
OSU_B=https://github.com/ppy/osu/pull/33961
SCORE_PROCESSOR_A=https://github.com/tsunyoku/osu-queue-score-statistics/tree/add-hd-so-to-relevant-legacy-mods
SCORE_PROCESSOR_B=https://github.com/tsunyoku/osu-queue-score-statistics/tree/add-hd-so-to-relevant-legacy-mods

Givikap120 commented:
> Target: #33961 Spreadsheet: https://docs.google.com/spreadsheets/d/1b-JVvUlKHvwyiszcefzjtyrLqG2qJ_FSwplMxasqunk/edit

That's weird, db doesn't have any lazer CL scores?

tsunyoku commented on osu.Game/Screens/Play/HUD/PerformancePointsCounter.cs:
```diff
@@ -56,7 +56,7 @@ private void load(BeatmapDifficultyCache difficultyCache)
                 clonedMods = gameplayState.Mods.Select(m => m.DeepClone()).ToArray();
 
-                scoreInfo = new ScoreInfo(gameplayState.Score.ScoreInfo.BeatmapInfo, gameplayState.Score.ScoreInfo.Ruleset) { Mods = clonedMods };
+                scoreInfo = new ScoreInfo(gameplayState.Score.ScoreInfo.BeatmapInfo, gameplayState.Score.ScoreInfo.Ruleset) { Mods = clonedMods, IsLegacyScore = gameplayState.Score.ScoreInfo.IsLegacyScore };
```
@ppy/team-client could I get an approval on this out-of-difficulty change? It sets `IsLegacyScore` so that we can distinguish between legacy scores and lazer CL scores in the PP calculator.

Givikap120 committed:
Add comment explaining lazer CL miss penalty

Givikap120 committed:
Update the comment

tsunyoku reviewed (changes requested):
It'd probably also be a good time to refactor this since we have more variables available to us here:

https://github.com/ppy/osu/blob/cf4d6bea72eefef1ca2e4e32fc483b50ee9f4686/osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs#L103-L114

We should be able to change the condition to simply `if (isLegacyScore)`, and opt for a `Debug.Assert(score.LegacyTotalScore != null)` or something like that. Although, I'm not sure if that would cause the PP counter to crash - please test, but this would read better I think.

Miss count logic is becoming a beast of its own, I think I want to split it further in the future if I can help it.

Givikap120 committed:
Update OsuPerformanceCalculator.cs

tsunyoku committed:
Remove confusing comment

tsunyoku reviewed (dismissed):
I don't really like the amount of moving parts that `calculateComboBasedEstimatedMissCount` has at this point but that's a problem for later.

Givikap120 committed:
Remove drain rate attribute (#35817)

Givikap120 committed:
Skip score-based miss count estimation for ScoreV2 (#35962)

* Add basic scorev2 support

* Revert unnecessary score statistic changes

* Fix CI

* Revert changes

* Disable score-based misscount for scoreV2

stanriders committed:
Use `DifficultyCalculationUtils.Norm` (#36064)

stanriders committed:
Simplify osu! star rating calculations (#36063)

* Simplify star rating calculations

* Refactor

Finadoggie committed:
Uses real distance in speed instead of mishandling variables (#35387)

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

Givikap120 committed:
change FL calc depending on HD settings (#33019)

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

Finadoggie committed:
Use number calculated at runtime from DecayWeight (#33733)

Replaces a constant that assumes DecayWeight == 0.9

Co-authored-by: James Wilson <tsunyoku@gmail.com>
Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

tsunyoku committed:
Reduce calls to `DifficultyValue` where possible (#36112)

stanriders committed:
Extend `Skill` to include `ObjectDifficulties` (#36152)

* Extend `Skill` to include `ObjectDifficulties`

* Remove generic

* Change `ObjectDifficulties` to be modifiable by children skills

* Fix tests

tsunyoku committed:
Add `HarmonicSkill` (#36153)

* Add `OsuHarmonicSkill`

* Make `ProcessInternal` a sealed override

* Move to main game project

stanriders committed:
Change the doubletapness scaling to be more punishing on the low end of the hitwindow (#36148)

Givikap120 committed:
Improve fullcombo threshold formula (#35724)

* Implement new portion formula

* Change the formula

* Use base value of 5 sliders

* Make the threshold harsher

---------

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

tsunyoku committed:
Fix legacy score base multiplier being calculated using modded difficulty values (#36209)

Fun correctness bug.

ScoreV1 multiplier uses the **nomod** peppy stars (i.e the beatmap's
"base" HP, OD and CS) in all cases, which was generally the intention
with this code but it passes the wrong beatmap resulting in it using
modded values for this.

Cross referenced with stable and this now results in the expected
multipliers.

Results in some very minor (<3pp from my testing) changes for HR and EZ
scores with combo breaks.

kwotaq committed:
Add comment for wiggle bonus multiplier (#36373)

It's a common mistake by newer pp dev contributors to increase the
wiggle multiplier not knowing its at its maximum, so I added a comment
to try to avoid more people doing it.

---------

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

Givikap120 committed:
Merge branch 'pp-dev' into use-lazer-info

Natelytle committed:
Move strain's influence in the difficulty of a note based on deltaTime to the evaluator level (#36417)

This PR moves the influence of d/t^2 from the skill (through strain)
directly into the evaluator level as a bonus applied at the end. This
makes it more clear what d/t^2 is, and the fact that it applies to *all*
bonuses in the evaluator. As well, the peak strain value of a note is
now equal to the evaluator value. This comes with a couple benefits:

1. The BPM weight is now subject to balance, and can be flattened easily
for a more "d/t" like system (previously this required hacky solutions)
2. StrainDecayBase becomes a much more useful variable now that it does
not affect the difficulty of the note itself. When adjusting this in
live, your star rating would double upon changing from 0.15 in aim to
0.3, and now it is intuitive what it does (makes strain take longer to
accumulate). This means that future balancing efforts can use evaluators
to dynamically adjust strainDecayBase (potentially letting large spikes
provide more strain for the same difficulty if they're wide angle, for
example).
3. In the object inspector, you get the actual maximum difficulty value
of a note as seen in the ObjectDifficulties list. This makes it easier
to tell what notes are deemed as harder by the system. For context,
previously, a note of difficulty 1 at 200bpm would cap out at 6 as a
result of strain, and a note of the same difficulty but at 300bpm would
cap out at 8.

The actual implementation is really really simple. I'm willing to move
this to flashlight if wanted (I don't think it's necessary), or even
abstract this away so that (1 - decay) doesn't look like a balancing
constant.

Side note: this is equivalent to live, except for notes with a deltaTime
of less than 25ms (since I am using AdjustedDeltaTime in the aim
evaluator for the bonus to avoid divisions by zero).

kwotaq committed:
Replace Speed skill strain summation with a per-note harmonic sum (#34696)

The purpose of this change is to remove the arbitrary note based length
bonus for speed and replace it with a more concrete per note difficulty
aware calculation. As a result of the summation needing to be per-note,
chunking and in consequence peak strain reduction have both been removed
from the skill. ~~The various strain counting functions also heavily
relied on chunking and the way summation worked, so they have also been
changed to match values held before the change as much as possible
(given the different summation some changes in values are bound to
happen).~~ Above was a bug.

Huis page:
[https://pp.huismetbenen.nl/rankings/players/length-bonus](https://pp.huismetbenen.nl/rankings/players/length-bonus)

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

nisico3 committed:
Improve angle calculation to better represent the path taken on sliders (#35555)

## This pr resolves the issue with the angle calculation.
The old system calculated the angle between three objects. If the last
object was a long slider that the player needed to follow, the system
calculated the angle using only the slider’s end point and the previous
object, which caused long sliders to have incorrect angle values.

---------

Co-authored-by: StanR <hi@stanr.info>
Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Use correct object for the aim slider bonus (#36216)

Currently the slider bonus works on the assumption that the travel
velocity of the previous slider is a part of the current object's
difficulty because it is part of the movement from prev to curr.
However, this is contradicted by the fact that `currVelocity` is a
combination of prev->curr + curr slider velocity and is actually
breaking the assumption that slider difficulty is contained in the
slider itself that we take when calculating difficult slider strains.

This makes it so that the slider bonus difficulty is contained in the
slider itself instead of buffing the next object, which makes both the
calculation overall more consistent and the slider factor calculation
actually work as expected.

Aim multiplier got slightly lowered because this change makes most of
the sliders gain a little bit

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

kwotaq committed:
Replace current AR and HD bonuses with a new Reading skill on the osu! ruleset (#33196)

This PR aims to replace the current bonuses used to award high approach
rates and scores made using the Hidden mod with a Reading skill that
takes into account each note's reading difficulty separately, Important
to note is the fact that as reading difficulty is now a skill the
bonuses are now additive instead of multiplicative, meaning there are
vast changes in deltas on the high and low end of scores. Due to the
nature of adding a new skill new difficulty and performance attributes
need to be added.

Huis page:
[https://pp.huismetbenen.nl/rankings/admin/kwotaq-reading](url)

---------

Co-authored-by: apollo-dw <83023433+apollo-dw@users.noreply.github.com>
Co-authored-by: js1086 <js1086@student.le.ac.uk>
Co-authored-by: tsunyoku <tsunyoku@gmail.com>
Co-authored-by: StanR <hi@stanr.info>

stanriders committed:
Move speed distance bonus to aim (#36168)

This doesn't solve _flow_ aim in any way, but makes it so that `Speed`
doesn't have distance scaling (read as "aim") anymore which fixes some
issues related to that like the length bonus behaving incorrectly, and
in general `Speed` being an aim+tap skill instead of just a tapping
skill

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Move TD difficulty reduction to Aim (#36476)

stanriders committed:
Use deviation-based Speed accuracy scaling (#36475)

https://pp.huismetbenen.nl/rankings/players/tap-statacc

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

Givikap120 committed:
Make reading and flashlight pp to be LP summed (#36464)

Right now they're summed normally what opens a problem that FL rewards
too much pp when combined with reading map, since you're memorizing FL
anyway.

Co-authored-by: James Wilson <tsunyoku@gmail.com>

Givikap120 committed:
Make sliderless aim in fact sliderless (#29993)

Part of this PR - https://github.com/ppy/osu/pull/27303

Current aim calculation have a flaw of sliderless aim still accounting
for sliders. This happens because of usage of `LazyJumpDistance` as a
main distance metric.

This PR is fixing this by adding `JumpDistance` as true sliderless
metric, using it instead of `LazyJumpDistance`.
This can introduce very rare cases where sliderless aim is worth more
than normal aim (because of velocity change bonus).

The effect of this is minimal on most of the maps. It can be seen the
best on this map - https://osu.ppy.sh/beatmapsets/594751#osu/1257904
Before:

![image](https://github.com/user-attachments/assets/e96773e6-3274-4ed6-8293-ed9bdfa213d0)

After:

![image](https://github.com/user-attachments/assets/126e861b-c28d-4ed4-a0db-f5305225a749)

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>
Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

stanriders committed:
Undo slider high cs bonus hotfix (#36513)

It existed for one map and now that that map is fine there's no need to
do it anymore.

Givikap120 committed:
Fix flashlight having near-zero bonus when reading difficulty is much higher (#36487)

There was a recent PR (https://github.com/ppy/osu/pull/36464) that was
aimed on accounting for the fact that getting your map partially
memorized with FL makes reading easier, so those bonuses shouldn't
reward full pp to each other.
But in cases where FL pp is significantly lower than reading it have
lead to cases where FL adds practically 0 additional pp.
This PR is adjusting a formula to account for cases like this.

FL reward on this map -
https://osu.ppy.sh/beatmapsets/1487999#osu/3259719 with EZHDHT(FL):
Full bonus: 408pp -> 486pp (+78pp)
Current: 408pp -> 424pp (+16pp)
This PR: 408pp -> 478pp (+70pp)

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

Givikap120 committed:
Adjust speed deviation scaling to be more harsh on low end (#36554)

This should buff raw speed plays like Ivaxa Violation, at the same time
undoing part of the buff on the lower end scores like Save Me NM
Moved part of the multiplier out of the Pow to be more intuitive (it
multiplies the 20 by Pow(difficulty/4), so it's more clear that it would
be equal to 1 on difficulty = 4)
The scaling itself was adjusted to be more similar to live (so
buffs/nerfs on 98% acc remains +- the same through the difficulty curve)

tsunyoku committed:
Nerf aim strain for objects <radius (#36558)

This should be a pretty simple fix for doubles being systemically broken
in aim. Doubles get essentially zero bonus from the aim eval itself -
wide & acute bonuses are zero due to the lack of distance, and velocity
is close to zero _again_ because of the lack of distance. However, the
delta times of these notes mean that `highBpmBonus` is very kind and
will buff the strain quite significantly.

This change means that the 2nd note of a double should get next to no
aim strain, which feels like correct behaviour. From testing, streams,
stacks etc. are essentially unchanged by this due to the fact its using
radius rather than diameter.

https://pp.huismetbenen.nl/rankings/players/doubles-strain

Givikap120 committed:
Simplify hidden difficulty calculation (#36465)

It removes unnecessary function `DurationSpentInvisible` function that
just rescaled preempt.
Now it's just using preempt directly.
I've made multiplier to be very close to the current one, so pp deltas
should be minimal.

stanriders committed:
Change wide bonus distance reduction to be consistent with `SpeedAim.SINGLE_SPACING_THRESHOLD` (#36573)

Slightly reducing aim/speedaim doubledipping, but mostly just done for
consistency and ease of understanding of existing relation between both

stanriders committed:
Add a bunch of rhythm improvements (#36569)

This does a couple of things:

- Adds slider->circle effective ratio adjustment
- Removes same polarity nerf for single delta count islands (which are
pretty much THE most unpredictable)
- Makes polarity nerf check actual deltas and not just the count
- Excludes spinners from ratio calculation

https://pp.huismetbenen.nl/rankings/players/minor-rhythm-improvements

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Add real sliderend check in rhythm (#36593)

Classic blunder. See
https://discord.com/channels/546120878908506119/619291828054655016/1468738477297504489

molneya committed:
only nerf total cognition sum when reading > flashlight (#36599)

Alternative to #36464 and #36487.

Reading and flashlight should be treated much more separately, since the
flashlight skill only accounts for object visibility and repetition, but
reading accounts for overlaps, high/low AR, etc. However, for certain
scores with very high reading (to the point of memorisation), the skill
indeed overlaps with flashlight.

This PR uses the normal strain sum exponent for reading and flashlight
(for player expectations), with an extra factor on flashlight to account
for the ease of adding flashlight to a partially memorised map when
reading is the dominant skill.

stanriders committed:
Return zero rhythm difficulty for `Swell` and `DrumRoll` (#36628)

Found in the middle of hitwindow usage refactor

stanriders committed:
Add early returns in `HarmonicSkill` and `OsuDifficultyCalculator` (#36644)

Givikap120 committed:
Nerf HD perfect stack reward (#36623)

Safety measure to prevent potential major abuse maps.
The alternative is reducing nerf on multiplier (for example 5000 instead
of 2500), but increase nerf if rhythm is the same.

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

stanriders committed:
Refactor `DifficultyHitObject` to include `ClockRate` and `HitWindows` (#36630)

This changes difficulty calculations to use `ClockRate` and `HitWindows`
from `DifficultyHitObjects` instead of providing them manually. It's
main purpose is unifying all the calculations to use the same source of
truth, and also allowing future work on supporting variable clockrates.

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Change accuracy pp to be uncapped (#36669)

This also transfers some of the accuracy pp length bonus to aim which
makes it at least somewhat difficulty-aware. Value changes are very mild
(10pp at most for most players)

<img width="1613" height="575" alt="image"
src="https://github.com/user-attachments/assets/f2d27230-6771-4c4b-932c-33a8d804002d"
/>

stanriders committed:
Adjust wide angle bonus to be closer to SpeedAim distance scaling (#36670)

stanriders committed:
Adjust RX-related multipliers (#36697)

Mostly just keeping RX sane. I don't think we want to go too deep into
the balancing here

stanriders committed:
Fix Aim's `skillMultiplierTotal` being applied incorrectly (#36699)

Fortunately its 1.0 so it didn't break anything

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Q1 2026 osu! Balancing pass (#36734)

Nerf mid-level snap aim, buff raw tapping, buff flow slightly, buff
highbpm aim slightly

https://pp.huismetbenen.nl/rankings/players/q1-2026-balancing

stanriders committed:
Use full hitwindow in taiko Rhythm skill (#36671)

stanriders committed:
Fix combined slider velocity calculation (#36773)

This changes how the current/previous velocity is being calculated in
Aim. Currently it's being calculated as an addition of 2 velocities
together which isn't really correct, instead here velocity is being
calculated as a (combined distances) / (combined times). On practice
this buffs underweighted slider maps while overweighted ones stay about
the same

https://pp.huismetbenen.nl/rankings/players/use-proper-slider-velocity

stanriders committed:
Aim strain influence reduction (#36792)

Currently, live aim evaluation prioritizes high bpms with lower
distances over low bpms with higher distances, something you may have
heard colloquially called "d/t^2" in here. This creates an inherent DT
advantage / NM & HR disadvantage in the average aim case.

This change makes it so the evaluation is nudged closer to the ideal
distance/time aim evaluation. In practice, higher spacing (distance)
and/or slower jumps should generally see a buff, lower spacing
(distance) and/or faster jumps should generally see a nerf. This
includes some 3-mod scores as HR spacing is underweighted regardless of
bpm.

It also affects streams the same way, where faster high bpm low spacing
streams (speedflow) are nerfed and high spacing low bpm streams are
buffed.

As a disclaimer: this isn't a complete fix, it is moreso a step in the
right direction (hence, a Nudge, per se) towards the ideal d/t world.
This also doesn't change anything angle-related so high-spacing farm
maps will gain pp - this will be addressed separately after this change.


https://pp.huismetbenen.nl/rankings/players/reduce-strain-influence

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Reduce strain influence a bit more (#36813)

Changes snap aim scaling to d/t^1.65 and speedflow scaling to d/t^1.9
(roughly). This is as much as we can nudge scaling towards d/t without
breaking everything

stanriders committed:
Adjust wide aim bonus cutoff after SpeedAim changes (#36835)

Removing exponent since after d/t rescaling it's just making low spaced
streams way too nerfed

stanriders committed:
Slightly increase TD penalty (#36855)

stanriders committed:
Increase high cs bonus in SpeedAim (#36856)

stanriders committed:
Add basic flow evaluation (#36902)

https://pp.huismetbenen.nl/rankings/players/stanr-aimsep

kwotaq committed:
Penalize angle repetition depending on the difference between vectors (#36559)

This is a slightly more intelligent angle repetition nerf that checks
the previous vectors to see if the pattern has any rotation or not, this
allows for harsh nerfs on current meta patterning like the N, X and V
patterns while keeping things like rotating 1-2s and triangles less
nerfed. The effect the vectors have on angle repetition is adjustable.

---------

Co-authored-by: StanR <hi@stanr.info>

peppy committed:
Merge branch 'master' into pp-dev

Rian8337 committed:
Fix Hidden mod perfect stacks buff for osu! reading difficulty (#36918)

The `OpacityAt` conditional adds `Preempt` to `StartTime` when the
intention is to check for when `previousObj` is hit, which means it
should just be `StartTime`.

In addition to that, the third conditional's purpose is to ensure that
the buff only applies if `currObj` is animating (its opacity is changing
from the Hidden mod) when `previousObj` is hit. This restructures it so
that it reads more cleanly and communicates its purpose better.


[Conversation](https://discord.com/channels/546120878908506119/1374004990875795586/1480886152042119190)

stanriders committed:
Remove `OsuStrainSkill` (#36917)

stanriders committed:
Organize osu! difficulty evaluators a bit better (#36921)

stanriders committed:
Adjust `Speed` harmonic summation params (#36923)

Slight balancing. Mostly exists to reduce extremely consistent tapping
maps (i.e [Age of
Tyranny](https://osu.ppy.sh/beatmapsets/2219231#osu/4704022)) influence
over the summation

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

tsunyoku committed:
Remove `OsuStrainUtils` (#36924)

The only method inside this class is used in a single place ever since
Speed became a harmonic skill so it doesn't really make sense to keep
around to me. As a result, supersedes #33295.

Also includes a fix to use `DecayWeight` for the consistent top strain,
exactly the same as what we're doing for `CountTopWeightedStrains`.

Rian8337 committed:
Use current object's `Preempt` when retrieving past visible objects in osu! reading difficulty (#36944)

In `retrievePastVisibleObjects`, the wrong object's `Preempt` is used to
determine the time where the currently assessed object becomes visible.
This PR fixes that issue.

Additionally, the visibility condition in
`retrieveCurrentVisibleObjectDensity` is rearranged to read similarly to
the one in `retrievePastVisibleObjects`.

This does not change current values as none of the mods that can
currently award pp modify an object's `TimePreempt` individually. The
only affected mod is Freeze Frame, which gains buffs across the board as
more objects are calculated to be visible now.

stanriders committed:
Restore precision buff in Agility (#36968)

Seems like it got lost somewhere in the #36902. Great upside is that
flow cs bonus can be reduced as it's supposed to be now since the bonus
is made with d/t^1.65 in mind and flow is d/t. Values-wise it's a small
precision buff

smoogipoo committed:
Merge branch 'master' into pp-dev

tsunyoku committed:
Apply a few minor fixes to flow aim evaluation (#36999)

Contains:

- A small bug fix from initial separation merge. Snap aim only applies
this bonus when slider travel distance is passed, so it should be the
same for flow. Cannot find any cases where values are affected, but in
theory the only difference this can make is slider factor calculations.
- Apply overlapping note factor to the acute bonus, so that direction
changes that overlap are awarded less. A small speed multiplier increase
is included to offset this
- Move angular velocity calculation out of ODHO into the flow evaluator,
and fix it referencing the wrong angle (2 objects back instead of the
previous object)
- 2 tiny code refactors

---------

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

stanriders committed:
Remove slider velocity considerations from previous object velocity (#36806)

It barely affects values at this point

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Adjust Aim RX multiplier (#36985)

Generally deflationary on the top end and roughly the same at low end.
Likely needs a more comprehensive solution, but this at least ensures
that even if we don't do it the values will be sane.

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Calculate clock time when creating difficulty hit objects (#36962)

This change removes `clockRate` precalculation from
`DifficultyCalculator`.

The idea is that clock rate should be calculated in-place (ideally for
every object) since we store and access it using DHOs. This also
prevents anyone from accidentally passing clock rate to skills

Unfortunately osu uses clock rate to calculate OD for the whole map in
`CreateDifficultyAttributes` so we can't make it completely DHO-based,
but I think one single in-place call to `ModUtils.CalculateRateWithMods`
in `CreateDifficultyAttributes` is fine

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Change the aim scaling slightly (#36967)

This slightly buffs aim on above ~7.5 stars maps and slightly nerfs it
below that

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Make miss penalty slightly harsher on first misses (#37040)

https://www.desmos.com/calculator/naggvbcz0a

This change makes first ~3 misses have harsher miss penalty, while 9+
misses get lighter penalty. 4-9 misscounts stay close to unchanged.

stanriders committed:
Undo Agility strain influence rescaling (#37081)

Now that agility isn't "speedaim" that covers both snap and flow we can
restore it back into it's original scaling. Arguably it should be _more_
than d/t^2 but that's a thing to explore separately

stanriders committed:
Adjust high CS bonuses after all the distance-time rescaling (#37088)

At this point maybe having a shared small circle bonus in ODHO isn't
even worth it since we have different d/t scalings in all evaluators.
Bonus was made for snap (~d/t^1.65), here it's adjusted to be higher on
agility (~d/t^2) and lower on flow (~d/t). Practically affects like [1
map](https://osu.ppy.sh/beatmapsets/2191876#osu/5192354)

stanriders committed:
Increase Aim strain decay base (#37099)

This buffs longer jump sections and slightly nerfs short jump spikes.
Mostly a way to slightly restore non-aimslop DT aim maps since they got
hit with all the nerfs a bit too hard

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

Givikap120 committed:
Account for alternating angles in reading angle factor (#36466)

This rework target is C-type nerf, as this map gets absurd amount of
reading pp, because angles are considered very unrepetitive. What
results in this score being worth absurd 1.4k pp.

This PR is up to heavy discussion because it can be made much more
general, touching more maps. So I wait on pp committee opinion on what
of the parts can be removed.

Current checks for angle to be nerfed:
- The smaller angle has to be very sharp: <20 degrees, full power on <5
degrees
- The larger angle has to be wide: >60 degrees, full power on >120
degrees

If pattern meets all the criteria - it would be considered repetitive.
For now practically no map meets this criteria to significant amount
except C-type.

Finadoggie committed:
Use variable-length strains to fix chunking-related issues (#33351)

- Closes https://github.com/ppy/osu/issues/25254

Solves chunking-related issues by starting a new chunk for every object,
and allowing chunks to be different lengths. This is retrofitted to
existing calculations, and doesn't have other ramifications in the way
something like tr3s's continuous strains do.

The effect of this can be seen in the video below. In live pp, the map
would lose sr when increasing the rate at certain points, but in the
video, the sr rises consistently as expected.


https://github.com/user-attachments/assets/2d58946a-9e0c-4f6b-912a-71dfe75c0f8a

A multiplier has been added inside `DifficultyValue()` to account for SR
being slightly lower from more granular summation.

When testing this rework in PerformanceCalculatorGUI, I recommend using
[this
fork](https://github.com/Finadoggie/osu-tools/tree/variable-length-strains).
I can't guarantee that the visuals it shows are correct, but they are
definitely more correct than not.

Edit: Use [this
fork](https://github.com/Finadoggie/osu-tools/tree/unsynced-variable-length-strains)
for testing now. Old one is now for testing
SynchronizedVariableLengthStrainSkill

---------

Co-authored-by: tsunyoku <tsunyoku@gmail.com>
Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

stanriders committed:
Very slight overall aim buff (#37134)

stanriders committed:
Aim balancing (#37147)

kwotaq committed:
Refactor Reading strain to match the rest of the skills (#37010)

This doesn't affect values as per the original refactor.

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

stanriders committed:
Move Aim mod-related difficulty adjustments to after the probability calculation (#37184)

Adjusting difficulty before we calculate the flow probability is
technically wrong. This makes flow adjustment on RX actually work as it
should instead of trying to lower the snap to achieve similar results.
TD flow buff was removed because here it actually made some maps _very_
buffed. Values are mostly the same otherwise

kwotaq committed:
Refactor SnapAimEvaluator bonuses for better readability (#37119)

This PR moves the addition of the bonuses closer to their declaration
and also renames some variables and functions to more closely resemble
the way they're used. The names can be discussed for I'm not sold on
them either.

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Fix snap angle repetition (#37286)

Classic blunder.

stanriders committed:
Fix low distance snap/flow relation (#37306)

Currently all low spaced (<radius) aim is evaluated as low difficulty
snap. After we separated snap, flow and agility we don't actually need
snap aim to be reduced for low distance - this is something that flow
aim should cover instead as low spacing is guaranteed to be either
flowed or not aimed at all. By extension this fixes the need for the
flow velocity bonus to be extremely high as it no longer clashes with
snap when the changes happen between low and high spacing.

piiidosu committed:
Rebalance TC Bonus (#36555)

stanriders committed:
Rescale wide snap aim bonus (#37411)

Low bpm wide angle aim nerf, high bpm wide angle aim buff

Givikap120 committed:
Use Mehs in the okAdjustment (#37264)

I'm not sure why they were excluded initially, so creating this PR as a
potential fix
If there's an actual reason for this then this PR is probably
unnecessary

Co-authored-by: James Wilson <tsunyoku@gmail.com>
Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

stanriders committed:
Remove flow acute bonus velocity sqrt (#37436)

It's not really needed anymore and the added benefit is that circular
alt gets slightly buffed

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

wulpine committed:
Nerf linear spacings in osu!catch (#37287)

This change targets patterns which have no spacing variability and
visually are linear. Such patterns don't require additional inputs and
you can walk/dash through them.

In particular, [this
map](https://osu.ppy.sh/beatmapsets/2485027#fruits/5454338) (ab)uses
them to the point it was temporarily unranked.

The curve for the nerf was chosen arbitrarily, it just seems to do the
job.

Co-authored-by: James Wilson <tsunyoku@gmail.com>

Natelytle committed:
Standardize difficulty variable naming (#37460)

All occurrences of a difficulty variable in an evaluator are now called
evalNameDifficulty, of the raw output of this in a skill are now
referred to as difficulty, and of the strain adjusted versions are
referred to as strain.

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

stanriders committed:
Move mod-based difficulty adjustments to skills (#37621)

First part of getting rid of the RatingCalculator. Values changes are
expected, but apart from TD every other mod adjustment is unranked so it
doesn't really matter much

stanriders committed:
Another batch of small rhythm evaluation fixes (#37609)

## [Reduce rhythm effective ratio for patterns that are speeding
up](https://github.com/ppy/osu/commit/56f66abf82b687d88e6fe83928d6dfa0f657b0a0)
Assuming the same ratio difficulty, speeding up rhythms are a bit easier
to play than slowing down ones

## [Reduce rhythm complexity sum if the final island is
long](https://github.com/ppy/osu/commit/912d81cbfedee286b634c4bcb9c4e286d91a3d88)
Currently due to how the rhythm complexity sum works rhythm difficulty
gets applied to most of the object's island - historical decay isn't
enough to counter difficulty carryover so we end up having non-zero
rhythm difficulty on long consistent rhythm patterns (for example if a
stream starts with an unusual ratio it will have non-zero rhythm
difficulty regardless of its length, even if its hundreds of objects
long). This applies a global difficulty reduction depending on the
current object's island length.
https://www.desmos.com/calculator/kvnpasbpt2

## [Fix islands being initialised incorrectly if the rhythm section is
longer than the historical
cutoff](https://github.com/ppy/osu/commit/ddf0fe758b691ca6e2a157fe335a41c4048e7e9b)

Due to the change above we now have to properly initialise islands at
the start of the rhythm loop - currently if the object we're evaluating
is a part of a >32 object consistent rhythm pattern (say a stream) it
never actually gets its island properly initialised. It doesn't matter
on the current system, but it does matter if we want to reduce
complexity sum using island length as uninitialised island always has
length of 1 so patterns that are >32 objects long stop getting their
complexity reduced after the 32th object

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

Givikap120 committed:
Add distance factor to anti-doubletap (#37636)

Circles that are spaced from each other can't be doubletapped, therefore
they were excluded from the low OD penalty in speed.

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

molneya committed:
Remove accuracy bonus when flashlight is enabled (#36900)

When flashlight and reading weren't their own separate skills, pp
bonuses were put into the aim/speed/accuracy values to make them worth
something over nomod. Now that flashlight and hidden are calculated as
their own skills, it doesn't make sense to provide a bonus because
accurate values are calculated in the skills themselves.

The global flashlight skill multiplier was increased to counteract the
slight nerf of removing this bonus.

Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

Givikap120 committed:
Make Ok adjustment formula more stable (#37263)

Current formula can output drastically different values on simlar
countOk / estimatedSliderBreaks amounts.
For example if countOk is 10, then changing estimatedSliderBreaks from 9
to 10 is gonna change okAdjustment from 0.15 to 0.05, what is a 3 times
lower resulting "estimated sliderbreaks" value.

This PR is fixing that in the most simple way - by increasing the buffer
constants, so very small variables won't skew the result as much.

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>
Co-authored-by: StanR <hi@stanr.info>
Co-authored-by: StanR <8269193+stanriders@users.noreply.github.com>

stanriders committed:
Refactor aim/speed sliderbreak estimation calculation calls (#37654)

buyaspacecube committed:
osu!taiko penalty for frequent rhythm changes with long gaps (#37200)

Co-authored-by: Jay Lawton <jaime@merkids.com>
Co-authored-by: James Wilson <tsunyoku@gmail.com>

smoogipoo committed:
Merge branch 'master' into pp-dev

stanriders committed:
Move remaining osu! rating adjustments to skills (#37623)

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Fix `sliderFactor` using incorrect aim rating conversion curve (#37624)

tsunyoku committed:
Update difficulty calculation tests (#37829)

Final changes are now locked in, so update tests ready for PR to
`master`.

Givikap120 committed:
Merge branch 'pp-dev' into use-lazer-info
